src/devices/bus/spectrum/plusd.cpp

```diff
@@ -0,0 +1,374 @@
+
+FLOPPY_FORMATS_MEMBER(spectrum_plusd_device::floppy_formats)
+	FLOPPY_MGT_FORMAT
+FLOPPY_FORMATS_END
+
+//-------------------------------------------------
+//  ROM( plusd )
+//-------------------------------------------------
+
+ROM_START(plusd)
+	ROM_REGION(0x2000, "rom", 0)
+	
+	ROM_DEFAULT_BIOS("gdos")
+	
+	ROM_SYSTEM_BIOS(0, "gdos", "G+DOS v1a")
+	ROMX_LOAD("plusd_g.rom", 0x0000, 0x2000, CRC(569f7e55) SHA1(6b841dc5797ef7eb219ad455cd1e434ca3b9d30d), ROM_BIOS(0))
+	
+	ROM_SYSTEM_BIOS(1, "unidos", "UNI-DOS v2")
+	ROMX_LOAD("plusd_uni.rom", 0x0000, 0x2000, CRC(60920496) SHA1(399c8c7c8335bc59849a2182c32347603fd0288a), ROM_BIOS(1))
+	
+	ROM_REGION( 0x200, "plds", 0 )
+	ROM_LOAD( "alice_pal20l8.ic4", 0x000, 0x144, CRC(60135856) SHA1(41273f13a3680b29ba84ae1e85829482c783c55e) )
+ROM_END
+
+//-------------------------------------------------
+//  device_add_mconfig - add device configuration
+//-------------------------------------------------
+
+void spectrum_plusd_device::device_add_mconfig(machine_config &config)
+{
+	WD1772(config, m_fdc, 16_MHz_XTAL / 2);
+	FLOPPY_CONNECTOR(config, "fdc:0", plusd_floppies, "35dd", spectrum_plusd_device::floppy_formats).enable_sound(true);
+	FLOPPY_CONNECTOR(config, "fdc:1", plusd_floppies, "35dd", spectrum_plusd_device::floppy_formats).enable_sound(true);
+	
+	/* printer port */
+	CENTRONICS(config, m_centronics, centronics_devices, "printer");
+	m_centronics->busy_handler().set(FUNC(spectrum_plusd_device::busy_w));
+	
+	/* software list */
+	SOFTWARE_LIST(config, "flop_list").set_original("spectrum_mgt_flop");
+}
+
+
+const tiny_rom_entry *spectrum_plusd_device::device_rom_region() const
+{
+	return ROM_NAME(plusd);
+}
+
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+//-------------------------------------------------
+//  spectrum_plusd_device - constructor
+//-------------------------------------------------
+
+spectrum_plusd_device::spectrum_plusd_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, SPECTRUM_PLUSD, tag, owner, clock)
+	, device_spectrum_expansion_interface(mconfig, *this)
+	, m_rom(*this, "rom")
+	, m_fdc(*this, "fdc")
+	, m_floppy(*this, "fdc:%u", 0)
+	, m_centronics(*this, "centronics")
+{
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void spectrum_plusd_device::device_start()
+{
+	memset(m_ram, 0, sizeof(m_ram));
+
+	save_item(NAME(m_romcs));
+	save_item(NAME(m_ram));
+	save_item(NAME(m_centronics_busy));
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void spectrum_plusd_device::device_reset()
+{
+	m_romcs = 0;
+	m_centronics_busy = false;
```

Comment:
This is a cached input state - you shouldn't clear it on device reset.  If the device attached to the Centronics port changes busy status in response to reset, you'll get a callback.  Clearing the flag on reset like this can get you out-of-sync.